consensus/service/src/timestamp_validator.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+use displaydoc::Display;
+
+/// Provides logic for validating and ensuring that a timestamp exists when
+/// combining `ConsensusValues`.
+
+const MAX_TIMESTAMP_AGE: u64 = 30 * 1000; // 30 seconds
+
+/// Errors related to validating timestamps
+#[derive(Debug, Display, Clone, PartialEq)]
+pub enum Error {
+    /// The timestamp is in the future now{0}, timestamp{1}
+    InFuture(u64, u64),
+    /** The timestamp is older than the last bock: last_bock timestamp{0},
+    timestamp{1} */
+    OlderThanLastBlock(u64, u64),
+    /// The timestamp is too far in the past now{0}, timestamp{1}
+    TooOld(u64, u64),
+    /// No timestamp provided
+    NoTimestamp,
+}
+
+pub fn validate(timestamp: u64, last_block_timestamp: u64) -> Result<(), Error> {
+    let now = std::time::SystemTime::now()
+        .duration_since(std::time::UNIX_EPOCH)
+        .unwrap()
+        .as_millis() as u64;
+
+    if timestamp <= last_block_timestamp {
+        return Err(Error::OlderThanLastBlock(last_block_timestamp, timestamp));
+    }
+
+    if timestamp > now {
+        return Err(Error::InFuture(now, timestamp));
```

Comment:
Is it possible that this will create issues if a timestamp comes from a remote node who is a few millisecond ahead of the current node for some reason?

Reply:
Conversation brought over to https://github.com/mobilecoinfoundation/mobilecoin/pull/3733/files#r1399786804